build.rs

```diff
@@ -134,52 +215,148 @@ fn fix_config_map(mut config_map: HashMap<String, String>) -> HashMap<String, St
     config_map
 }
 
-fn load_cross_compile_info() -> Result<(InterpreterConfig, HashMap<String, String>)> {
-    let python_include_dir = env::var("PYO3_CROSS_INCLUDE_DIR")?;
-    let python_include_dir = Path::new(&python_include_dir);
+fn parse_script_output(output: &str) -> HashMap<String, String> {
+    output
+        .lines()
+        .filter_map(|line| {
+            let mut i = line.splitn(2, ' ');
+            Some((i.next()?.into(), i.next()?.into()))
+        })
+        .collect()
+}
 
-    let patchlevel_defines = parse_header_defines(python_include_dir.join("patchlevel.h"))?;
+/// Parse sysconfigdata file
+///
+/// The sysconfigdata is basically a dictionary, and since we can't really use this library to read
+/// it for us, egg and chicken type thing, we parse it with some regex. Here there are two regex's.
+/// The first one is to match an entry in the dictionary (`entry_re`). Then when the entry is on
+/// multiple lines we use the second regex to capture the additional string entry and add it to the
+/// previously captured key. We detect if this is a multi line entry by checking if the last capture
+/// group contains either `,` or `}`.
+fn parse_sysconfigdata(config_path: impl AsRef<Path>) -> Result<HashMap<String, String>> {
+    let mut script = fs::read_to_string(config_path)?;
+    script += r#"
+print("version_major", build_time_vars["VERSION"][0])  # 3
+print("version_minor", build_time_vars["VERSION"][2])  # E.g., 8
+if "WITH_THREAD" in build_time_vars:
+    print("WITH_THREAD", build_time_vars["WITH_THREAD"])
+if "Py_TRACE_REFS" in build_time_vars:
+    print("Py_TRACE_REFS", build_time_vars["Py_TRACE_REFS"])
+if "COUNT_ALLOCS" in build_time_vars:
+    print("COUNT_ALLOCS", build_time_vars["COUNT_ALLOCS"])
+if "Py_REF_DEBUG" in build_time_vars:
+    print("Py_REF_DEBUG", build_time_vars["Py_REF_DEBUG"])
+print("Py_DEBUG", build_time_vars["Py_DEBUG"])
+print("Py_ENABLE_SHARED", build_time_vars["Py_ENABLE_SHARED"])
+print("LDVERSION", build_time_vars["LDVERSION"])
+print("SIZEOF_VOID_P", build_time_vars["SIZEOF_VOID_P"])
+"#;
+    let output = run_python_script(&find_interpreter()?, &script)?;
+
+    Ok(parse_script_output(&output))
+}
+
+fn starts_with(entry: &DirEntry, pat: &str) -> bool {
+    let name = entry.file_name();
+    name.to_string_lossy().starts_with(pat)
+}
+fn ends_with(entry: &DirEntry, pat: &str) -> bool {
+    let name = entry.file_name();
+    name.to_string_lossy().ends_with(pat)
+}
 
-    let major = match patchlevel_defines
-        .get("PY_MAJOR_VERSION")
-        .map(|major| major.parse::<u8>())
-    {
-        Some(Ok(major)) => major,
-        Some(Err(e)) => bail!("Failed to parse PY_MAJOR_VERSION: {}", e),
-        None => bail!("PY_MAJOR_VERSION undefined"),
+fn find_sysconfigdata(path: impl AsRef<Path>, cross: &CrossCompileConfig) -> Option<PathBuf> {
+    for f in fs::read_dir(path).expect("Path does not exist") {
+        return match f {
```

Comment:
I think it's better we have some comments that describe why this complicated match is necessary.
[Your issue comment](https://github.com/PyO3/pyo3/pull/1095#issuecomment-674976471) is nice and copy-pasting a part of it is sufficient, though.